server/src/main/java/org/apache/druid/client/DirectDruidClient.java

```diff
@@ -231,6 +234,11 @@ private boolean enqueue(ChannelBuffer buffer, long chunkNum) throws InterruptedE
           final long currentQueuedByteCount = queuedByteCount.addAndGet(holder.getLength());
           queue.put(holder);
 
+          if (usingBackpressure && currentQueuedByteCount >= maxQueuedBytes && channelReadable.get()) {
```

Comment:
There's a race with how channelReadable is handled here: enqueue and dequeue are called from different threads and can run simultaneously. In fact, dequeue can get a chunk even before enqueue is finished running (dequeue is eligible to get a chunk as soon as the queue.put call is made in enqueue).

I think it'd be better for NettyHttpClient to track the time that reads were suspended, and return that from TrafficCop.resume, instead of returning void like it does now. Then DirectDruidClient only has to use that value to increment channelSuspendedTime.

Reply:
Btw, the reason I suggest this is that NettyHttpClient already has some synchronization around adjusting readability of the channel. Check out the logic protected by `watermarkLock`.

Reply:
@gianm Thanks so much for the review. Your feedback was really useful and I've made the changes that you requested.

Reply:
Also, tests seem to be passing -- not sure why these warnings occur however: https://travis-ci.org/apache/incubator-druid/builds/432030454?utm_source=github_status&utm_medium=notification